contrib/sort/sort.1
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" %sccs.include.redist.roff%
.\"
.\"     @(#)sort.1	8.1 (Berkeley) 06/06/93
.\"
.Dd 
.Dt SORT 1
.Os
.Sh NAME
.Nm sort
.Nd sort or merge text files
.Sh SYNOPSIS
.Nm sort
.Op Fl cmubdfinr
.Op Fl t Ar char
.Op Fl T Ar char
.Oo
.Cm Fl k Ar field1[,field2]
.Oc
.Ar ...
.Op Fl o Ar output
.Op Ar file
.Ar ...
.Sh DESCRIPTION
The
.Nm sort
utility
sorts text files by lines.
Comparisons are based on one or more sort keys extracted
from each line of input, and are performed
lexicographically. By default, if keys are not given,
.Nm sort
regards each input line as a single field.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl c
Check that the single input file is sorted.
If the file is not sorted,
.Nm sort
produces the appropriate error messages and exits with code 1;
otherwise,
.Nm sort
returns 0.
.Nm Sort
.Fl c
produces no output.
.It Fl m
Merge only; the input files are assumed to be pre-sorted.
.It Fl o Ar output
The argument given is the name of an
.Ar output
file to
be used instead of the standard output.
This file
can be the same as one of the input files.
.It Fl u
Unique: suppress all but one in each set of lines
having equal keys.
If used with the
.Fl c
option,
check that there are no lines with duplicate keys.
.El
.Pp
The following options override the default ordering rules.
When ordering options appear independent of key field
specifications, the requested field ordering rules are
applied globally to all sort keys.
When attached to a specific key (see
.Fl k ) ,
the ordering options override
all global ordering options for that key.
.Bl -tag -width indent
.It Fl d
Only blank space and alphanumeric characters
.\" according
.\" to the current setting of LC_CTYPE
are used
in making comparisons.
.It Fl f
Considers all lowercase characters that have uppercase
equivalents to be the same for purposes of
comparison.
.It Fl i
Ignore all non-printable characters.
.It Fl n
An initial numeric string, consisting of optional
blank space, optional minus sign, and zero or more
digits (including decimal point)
.\" with
.\" optional radix character and thousands
.\" separator
.\" (as defined in the current locale),
is sorted by arithmetic value.
(The
.Fl n
option no longer implies
the
.Fl b
option.)
.It Fl r
Reverse the sense of comparisons.
.El
.Pp
The treatment of field separators can be altered using the
options:
.Bl -tag -width indent
.It Fl b
Ignores leading blank space when determining the start
and end of a restricted sort key.
A
.Fl b
option specified before the first
.Fl k
option applies globally to all
.Fl k
options.
Otherwise, the
.Fl b
option can be
attached independently to each
.Ar field
argument of the
.Fl k
option (see below).
Note that the
.Fl b
option
has no effect unless key fields are specified.
.It Fl t Ar char
.Ar Char
is used as the field separator character. The initial
.Ar char
is not considered to be part of a field when determining
key offsets (see below).
Each occurrence of
.Ar char
is significant (for example,
.Dq Ar charchar
delimits an empty field).
If
.Fl t
is not specified,
blank space characters are used as default field
separators.
.It Fl T Ar char
.Ar Char
is used as the record separator character.
This should be used with discretion;
.Fl T Ar <alphanumeric>
usually produces undesirable results.
The default line separator is newline.
.It Fl k Ar field1[,field2]
Designates the starting position,
.Ar field1 ,
and optional ending position, 
.Ar field2 ,
of a key field.
The
.Fl k
option replaces the obsolescent options
.Cm \(pl Ns Ar pos1
and
.Fl Ns Ar pos2 .
.El
.Pp
The following operands are available:
.Bl -tag -width indent
.Ar file
The pathname of a file to be sorted, merged, or checked.
If no file
operands are specified, or if
a file operand is
.Fl ,
the standard input is used.
.Pp
A field is
defined as a minimal sequence of characters followed by a
field separator or a newline character.
By default, the first
blank space of a sequence of blank spaces acts as the field separator.
All blank spaces in a sequence of blank spaces are considered
as part of the next field; for example, all blank spaces at
the beginning of a line are considered to be part of the
first field.
.Pp
Fields are specified
by the
.Fl k Ar field1[,field2]
argument. A missing
.Ar field2
argument defaults to the end of a line.
.Pp
The arguments
.Ar field1
and
.Ar field2
have the form
.Em m.n
followed by one or more of the options
.Fl b , d , f , i ,
.Fl n , r .
A
.Ar field1
position specified by
.Em m.n
.Em (m,n > 0)
is interpreted as the
.Em n Ns th
character in the
.Em m Ns th
field.
A missing
.Em \&.n
in
.Ar field1
means
.Ql \&.1 ,
indicating the first character of the
.Em m Ns th
field;
If the
.Fl b
option is in effect,
.Em n
is counted from the first
non-blank character in the
.Em m Ns th
field;
.Em m Ns \&.1b
refers to the first
non-blank character in the
.Em m Ns th
field.
.Pp
A
.Ar field2
position specified by
.Em m.n
is interpreted as
the
.Em n Ns th
character (including separators) of the
.Em m Ns th
field.
A missing
.Em \&.n
indicates the last character of the 
.Em m Ns th
field;
.Em m 
= \&0
designates the end of a line.
Thus the option
.Fl k Ar v.x,w.y
is synonymous with the obsolescent option
.Cm \(pl Ns Ar v-\&1.x-\&1
.Fl Ns Ar w-\&1.y ;
when
.Em y
is omitted,
.Fl k Ar v.x,w
is synonymous with
.Cm \(pl Ns Ar v-\&1.x-\&1 
.Fl Ns Ar w+1.0 .
The obsolescent
.Cm \(pl Ns Ar pos1
.Fl Ns Ar pos2
option is still supported, except for
.Fl Ns Ar w\&.0b,
which has no
.Fl k
equivalent.
.Sh FILES
.Bl -tag -width Pa -compact
.It Pa /var/tmp/sort.*
Default temporary directories.
.It Pa Ar output Ns #PID
Temporary name for
.Ar output
if
.Ar output
already exists.
.El
.Sh SEE ALSO
.Xr comm 1 ,
.Xr uniq 1 ,
.Xr join 1
.Sh RETURN VALUES
Sort exits with one of the following values:
.Bl -tag -width flag -compact
.It Pa 0:
normal behavior.
.It Pa 1:
on disorder (or non-uniqueness) with the
.Fl c
option
.It Pa 2:
an error occurred.
.Sh BUGS
Lines longer than 65522 characters are discarded and processing continues.
To sort files larger than 60Mb, use
.Nm sort
.Fl H ;
files larger than 704Mb must be sorted in smaller pieces, then merged.
To protect data
.Nm sort
.Fl o
calls link and unlink, and thus fails in protected directories.
.Sh HISTORY
A
.Nm sort
command appeared in
.At v6 .
.Sh NOTES
The current sort command uses lexicographic radix sorting, which requires
that sort keys be kept in memory (as opposed to previous versions which used quick
and merge sorts and did not.)
Thus performance depends highly on efficient choice of sort keys, and the
.Fl b
option and the
.Ar field2
argument of the
.Fl k
option should be used whenever possible.
Similarly,
.Nm sort
.Fl k1f
is equivalent to
.Nm sort
.Fl f
and may take twice as long.


contrib/sort/sort.c
/*-
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Peter McIlroy.
 *
 * %sccs.include.redist.c%
 */

#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)sort.c	8.1 (Berkeley) 06/06/93";
#endif /* not lint */

/* Sort sorts a file using an optional user-defined key.
 * Sort uses radix sort for internal sorting, and allows
 * a choice of merge sort and radix sort for external sorting.
 */

#include "sort.h"
#include "fsort.h"
#include "pathnames.h"

#include <paths.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

int REC_D = '\n';
u_char d_mask[NBINS];		/* flags for rec_d, field_d, <blank> */
/*
 * weight tables.  Gweights is one of ascii, Rascii..
 * modified to weight rec_d = 0 (or 255)
 */
extern u_char gweights[NBINS];
u_char ascii[NBINS], Rascii[NBINS], RFtable[NBINS], Ftable[NBINS];
/*
 * masks of ignored characters.  Alltable is 256 ones
 */
u_char dtable[NBINS], itable[NBINS], alltable[NBINS];
int SINGL_FLD = 0, SEP_FLAG = 0, UNIQUE = 0;
struct coldesc clist[(ND+1)*2];
int ncols = 0;
extern struct coldesc clist[(ND+1)*2];
extern int ncols;

char devstdin[] = _PATH_STDIN;
char toutpath[_POSIX_PATH_MAX];

static void cleanup __P((void));
static void onsig __P((int));
static void usage __P((char *));

int
main(argc, argv)
	int argc;
	char *argv[];
{
	extern int optind;
	extern char *optarg;
	int (*get)();
	int ch, i, stdinflag = 0, tmp = 0;
	char cflag = 0, mflag = 0, nflag = 0;
	char *outfile, *outpath = 0;
	struct field fldtab[ND+2], *ftpos;
	union f_handle filelist;
	FILE *outfd;
	memset(fldtab, 0, (ND+2)*sizeof(struct field));
	memset(d_mask, 0, NBINS);
	d_mask[REC_D = '\n'] = REC_D_F;
	SINGL_FLD = SEP_FLAG = 0;
	d_mask['\t'] = d_mask[' '] = BLANK | FLD_D;
	ftpos = fldtab;
	fixit(&argc, argv);
	while ((ch = getopt(argc, argv, "bcdfik:mHno:rt:T:ux")) != EOF) {
	switch (ch) {
		case 'b': fldtab->flags |= BI | BT;
			break;
		case 'd':
		case 'i':
		case 'f':
		case 'r': tmp |= optval(ch, 0);
			if (tmp & R && tmp & F)
				fldtab->weights = RFtable;
			else if (tmp & F)
				fldtab->weights = Ftable;
			else if(tmp & R)
				fldtab->weights = Rascii;
			fldtab->flags |= tmp;
			break;
		case 'o':
			outpath = optarg;
			break;
		case 'n':
			nflag = 1;
			setfield("1n", ++ftpos, fldtab->flags&(~R));
			break;
		case 'k':
			 setfield(optarg, ++ftpos, fldtab->flags);
			break;
		case 't':
			if (SEP_FLAG)
				usage("multiple field delimiters");
			SEP_FLAG = 1;
			d_mask[' '] &= ~FLD_D;
			d_mask['\t'] &= ~FLD_D;
			d_mask[*optarg] |= FLD_D;
			if (d_mask[*optarg] & REC_D_F)
				err(2, "record/field delimiter clash");
			break;
		case 'T':
			if (REC_D != '\n')
				usage("multiple record delimiters");
			if ('\n' == (REC_D = *optarg))
				break;
			d_mask['\n'] = d_mask[' '];
			d_mask[REC_D] = REC_D_F;
			break;
		case 'u':
			UNIQUE = 1;
			break;
		case 'c':
			cflag = 1;
			break;
		case 'm':
			mflag = 1;
			break;
		case 'H':
			PANIC = 0;
			break;
		case '?':
		default: usage("");
		}
	}
	if (cflag && argc > optind+1)
		errx(2, "too many input files for -c option");
	if (argc - 2 > optind && !strcmp(argv[argc-2], "-o")) {
		outpath = argv[argc-1];
		argc -= 2;
	}
	if (mflag && argc - optind > (MAXFCT - (16+1))*16)
		errx(2, "too many input files for -m option");
	for (i = optind; i < argc; i++) {
		/* allow one occurrence of /dev/stdin */
		if (!strcmp(argv[i], "-") || !strcmp(argv[i], devstdin)) {
			if (stdinflag)
				warnx("ignoring extra \"%s\" in file list",
				    argv[i]);
			else {
				stdinflag = 1;
				argv[i] = devstdin;
			}
		} else if (ch = access(argv[i], R_OK))
			err(2, "%s", argv[i]);
	}
	if (!(fldtab->flags & (I|D) || fldtab[1].icol.num)) {
		SINGL_FLD = 1;
		fldtab[0].icol.num = 1;
	} else {
		if (!fldtab[1].icol.num) {
			fldtab[0].flags &= ~(BI|BT);
			setfield("1", ++ftpos, fldtab->flags);
		}
		if (nflag)
			fldtab[1].flags |= fldtab->flags;
		fldreset(fldtab);
		fldtab[0].flags &= ~F;
	}
	settables(fldtab[0].flags);
	num_init();
	fldtab->weights = gweights;
	if (optind == argc)
		argv[--optind] = devstdin;
	filelist.names = argv+optind;
	if (SINGL_FLD)
		get = makeline;
	else
		get = makekey;
	if (cflag) {
		order(filelist, get, fldtab);
		/* NOT REACHED */
	}
	if (!outpath) {
		(void)snprintf(toutpath,
		    sizeof(toutpath), "%sstdout", _PATH_DEV);
		outfile = outpath = toutpath;
	} else if (!(ch = access(outpath, 0)) &&
	    strncmp(_PATH_DEV, outpath, 5)) {
		struct sigaction act = {0, SIG_BLOCK, 6};
		int sigtable[] = {SIGHUP, SIGINT, SIGPIPE, SIGXCPU, SIGXFSZ,
		    SIGVTALRM, SIGPROF, 0};
		errno = 0;
		if (access(outpath, W_OK))
			err(2, "%s", outpath);
		act.sa_handler = cleanup;
		(void)snprintf(toutpath, sizeof(toutpath), "%sXXXX", outpath);
		outfile = mktemp(toutpath);
		if (!outfile)
			err(2, "%s", toutpath);
		(void)atexit(cleanup);
		for (i = 0; sigtable[i]; ++i)	/* always unlink toutpath */
			sigaction(sigtable[i], &act, 0);
	} else outfile = outpath;
	if (!(outfd = fopen(outfile, "w")))
		err(2, "%s", outfile);
	if (mflag)
		fmerge(-1, filelist, argc-optind, get, outfd, putline, fldtab);
	else
		fsort(-1, 0, filelist, argc-optind, outfd, fldtab);
	if (outfile != outpath) {
		if (access(outfile, 0))
			err(2, "%s", outfile);
		(void)unlink(outpath);
		if (link(outfile, outpath))
			err(2, "cannot link %s: output left in %s",
			    outpath, outfile);
		(void)unlink(outfile);
	}
	exit(0);
}

static void
onsig(s)
	int s;
{
	cleanup();
	exit(2);			/* return 2 on error/interrupt */
}

static void
cleanup()
{
	if (toutpath[0])
		(void)unlink(toutpath);
}

static void
usage(msg)
	char *msg;
{
	if (msg)
		(void)fprintf(stderr, "sort: %s\n", msg);
	(void)fprintf(stderr, "usage: [-o output] [-cmubdfinr] [-t char] ");
	(void)fprintf(stderr, "[-T char] [-k keydef] ... [files]\n");
	exit(2);
}
